content/questions/Recursion with slice function/index.md

```diff
@@ -0,0 +1,29 @@
+---
+title: Recursion with slice()
+tags:
+- recursion
+- slice
+order: 17
+date: '2019-10-05'
+answers:
+- 'Hello World'
+- 'e'
+- 'H // correct'
+- 'd'
+---
+
+Consider the following recursive function. If we pass the string `"Hello World"` to it, what gets logged?
+
+```javascript
+var speak = str => {
+
+if (str.length > 1) {
+return speak(str.slice(0,1));
+}
+return str;
+};
+console.log(speak('Hello world'));
+```
+<!---explanation--->
+
+The first time we call the function, `str.length` is greater than 1. We return the same function called on `str.slice(0,1)`, which is the string " "(that means, starting from no string). We repeat this process until the string reach the first character: the character "H". We then log that character.
```

Comment:
I think the second sentence needs to be "We return the same function called on `str.slice(0, -1)`", which is the string "Hello worl".

Reply:
Actually, in your case, the characters are reducing from right to left. ` str.slice(0, -1) `.
But in the above question, the characters are counting from left to right. i.e., 0 to 1.💁🏽‍♂️
Thus it gives ' H ' 😇

Reply:
What does `which is the string " "(that means, starting from no string)` mean then? I don't understand the explanation.

Currently, your recursive function will do this:

- "Hello World"
- "H"

and that's it. Was that the intent?

Reply:
It shows two things:
🔸`if (str.length > 1)` gives the number of characters to be counted.
🔸In `(str.slice (0,1))' , we starts before the letter H and counts 1 letter, that is H.

Reply:
If we will increase the number in `(str.length > 2) ` , it will give ` He `

Reply:
The intent is to print one letter and only the specified position of character, at the same time. 😀

Reply:
`if (str.length > 2)` will still print "H" because `str.slice(0,1)` will be just the first character.

Reply:
![image](https://user-images.githubusercontent.com/45743584/66440853-0db9a980-ea52-11e9-8c2e-d6b955b3bcd5.jpeg)

Reply:
![image](https://user-images.githubusercontent.com/45743584/66440867-232ed380-ea52-11e9-87e3-ada2a26f78fb.jpeg)

Reply:
So, the intent is to print required number of characters from specified position.
Thus, ` if (str.length > 2) ` can give MAXIMUM of two characters.
And   `  if (str.length > 1) ` can give MAXIMUM of one character. 🤷🏽‍♂️😇